src/charm.py

```diff
@@ -424,6 +453,83 @@ def _on_config_changed(self, event: ConfigChangedEvent) -> None:
+            direct_mongo.set_user_password(
+                OperatorUser.get_username(), direct_mongo.config.password
+            )
+        # Operator user not present in users, need to create it.
+        elif self.unit.is_leader():
+            logger.info("[Recovery] Create missing operator user (leader).")
+            direct_mongo.create_user(self.mongodb_config)
+        # Not leader and operator present, drop it.
+        elif OperatorUser.get_username() in users:
+            logger.info("[Recovering] Dropping operator user (not leader).")
+            direct_mongo.drop_user(OperatorUser.get_username())
+        # Drop backup and monitor user if present.
+        logger.info("[Recovering] Dropping users backup and monitor.")
+        if BackupUser.get_username() in users:
+            direct_mongo.drop_user(BackupUser.get_username())
+        if MonitorUser.get_username() in users:
+            direct_mongo.drop_user(MonitorUser.get_username())
+        pass
+
+    def _fix_mongodb_for_reuse(self) -> None:
+        """If we are reusing storage, we need to fix multiple things in the mongodb data.
+
+        This is done before starting (so with a downtime).
+        We need to:
+         * Start mongod in degraded mode (no auth validation, no replicaset
+         specified, bind only to local IP).
+         * Update operator password (on leader) and delete it on other units.
+         * Remove backup and monitor users on all units.
+         * Delete the local database (contains the replica set local information) on all units.
+        """
+        logger.debug("Fixing MongoDB for reuse.")
+        self.status.set_and_share_status(MaintenanceStatus("Updating MongoDB for storage reuse."))
+        try:
+            # Update with degraded configuration.
+            # Dangerous, this is only allowed here because we need to fix the DB configuration.
+            update_mongod_service(
+                machine_ip="127.0.0.1",
+                config=self.mongodb_config,
+                role=self.role,
+                degraded=True,
+            )
+            # Start the charm services in degraded mode.
+            self.start_charm_services()
+            # Perform data modification
+            if self.unit.is_leader():
+                # We can create only if we are the leader.
+                self._check_or_set_user_password(OperatorUser)
+            with MongoDBConnection(
+                self.mongodb_config, uri="localhost", direct=True
+            ) as direct_mongo:
+                for attempt in Retrying(
+                    stop=stop_after_attempt(10),
+                    wait=wait_fixed(5),
+                    reraise=True,
+                    before=before_log(logger, logging.DEBUG),
+                ):
+                    with attempt:
+                        if not direct_mongo.is_ready:
+                            raise NotReadyError
+                self.__fix_users_for_reuse(direct_mongo)
```

Comment:
Yes it is otherwise the users will not be propagated and patched on other nodes.
I tried to avoid doing this but it would end up in failures because credentials were not updated on other nodes.